include/ua_plugin_historical_access.h

```diff
@@ -0,0 +1,149 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. 
+ *
+ *    Copyright 2018 (c) basysKom GmbH <opensource@basyskom.com> (Author: Peter Rustler)
+ *    Copyright 2018 (c) Fraunhofer IOSB (Author: Julius Pfrommer)
+ */
+
+#ifndef UA_PLUGIN_HISTORICAL_ACCESS_H_
+#define UA_PLUGIN_HISTORICAL_ACCESS_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "ua_server.h"
+
+/**
+ * .. _historical-access:
+ *
+ * Historical Access Plugin API
+ * ============================
+ *
+ * Storage and access to historical data is implemented in a user-defined
+ * plugin. The plugin API does not mandate a specific kind of storage backend or
+ * handling of continuation points.
+ *
+ * The plugin is set globally in the server and used by the HistoryRead and
+ * HistoryUpdate services.
+ *
+ * The function pointers in the HistoryAccess plugin can be set to NULL. The
+ * server returns ``UA_STATUSCODE_BADNOTIMPLEMENTED`` if a requested method is
+ * not available.
+ */
+
+typedef struct {
+    void *context;
+    void (*deleteMembers)(void *context);
+
+    /* Is the node historizing? The callback is used by the normal Read service
+     * for the historizing attribute. */
+    UA_Boolean (*isHistorizing)(UA_Server *server, void *haContext,
+                                const UA_NodeId *nodeId);
+
+    /***************/
+    /* HistoryRead */
+    /***************/
+
+    UA_StatusCode
+    (*historyRead_raw)(UA_Server *server, void *haContext,
+                       const UA_NodeId *sessionId, void *sessionContext,
+                       const UA_ReadRawModifiedDetails *details,
+                       const UA_HistoryReadValueId *id,
```

Comment:
The request can consist of several nodes, as "this parameter contains the list of items upon which the historical retrieval is to be performed" (OPC UA Spec Part 4). How to determine, how many items are requested here?

Reply:
The demuxing would be done by the server before dropping into the plugin.

See the definition of the data types coming in.

Reply:
This could cause problems with some databases. Time critical, high performance or fail proof hardware often writes data as soon as possible to the storage, to have recent data in place after a fail occurred.

As they might write the data of multiple channels to the same files, you would need to iterate over the same files / time ranges for every channel, which destroys performance. You could collect the data of all channels in one go, if you know what to look for.